precompiles/assets-erc20/src/eip2612.rs

```diff
@@ -0,0 +1,275 @@
+impl_prefix!(Instance6, "Erc20Instance6Assets");
+impl_prefix!(Instance7, "Erc20Instance7Assets");
+impl_prefix!(Instance8, "Erc20Instance8Assets");
+impl_prefix!(Instance9, "Erc20Instance9Assets");
+impl_prefix!(Instance10, "Erc20Instance10Assets");
+impl_prefix!(Instance11, "Erc20Instance11Assets");
+impl_prefix!(Instance12, "Erc20Instance12Assets");
+impl_prefix!(Instance13, "Erc20Instance13Assets");
+impl_prefix!(Instance14, "Erc20Instance14Assets");
+impl_prefix!(Instance15, "Erc20Instance15Assets");
+impl_prefix!(Instance16, "Erc20Instance16Assets");
+
+/// Storage type used to store EIP2612 nonces.
+pub type NoncesStorage<Instance> = StorageDoubleMap<
+	<Instance as InstanceToPrefix>::NoncesPrefix,
+	Blake2_128Concat,
+	// Asset contract address
+	H160,
+	Blake2_128Concat,
+	// Owner
+	H160,
+	// Nonce
+	U256,
+	ValueQuery,
+>;
+
+pub struct Eip2612<Runtime, IsLocal, Instance: 'static = ()>(
+	PhantomData<(Runtime, IsLocal, Instance)>,
+);
+
+impl<Runtime, IsLocal, Instance> Eip2612<Runtime, IsLocal, Instance>
+where
+	Instance: InstanceToPrefix + 'static,
+	Runtime: pallet_assets::Config<Instance>
+		+ pallet_evm::Config
+		+ frame_system::Config
+		+ pallet_timestamp::Config,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_assets::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	BalanceOf<Runtime, Instance>: TryFrom<U256> + Into<U256> + EvmData,
+	Runtime: AccountIdAssetIdConversion<Runtime::AccountId, AssetIdOf<Runtime, Instance>>,
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin: OriginTrait,
+	IsLocal: Get<bool>,
+	<Runtime as pallet_timestamp::Config>::Moment: Into<U256>,
+{
+	fn compute_domain_separator(address: H160, asset_id: AssetIdOf<Runtime, Instance>) -> [u8; 32] {
+		let name: H256 =
+			keccak_256(pallet_assets::Pallet::<Runtime, Instance>::name(asset_id).as_slice())
+				.into();
+
+		let version: H256 = keccak256!("1").into();
+
+		let chain_id: U256 = Runtime::ChainId::get().into();
+
+		let domain_separator_inner = EvmDataWriter::new()
+			.write(H256::from(PERMIT_DOMAIN))
+			.write(name)
+			.write(version)
+			.write(chain_id)
+			.write(Address(address))
+			.build();
+
+		keccak_256(&domain_separator_inner).into()
+	}
+
+	pub fn generate_permit(
+		address: H160,
+		asset_id: AssetIdOf<Runtime, Instance>,
+		owner: H160,
+		spender: H160,
+		value: U256,
+		nonce: U256,
+		deadline: U256,
+	) -> [u8; 32] {
+		let domain_separator = Self::compute_domain_separator(address, asset_id);
+
+		let permit_content = EvmDataWriter::new()
+			.write(H256::from(PERMIT_TYPEHASH))
+			.write(Address(owner))
+			.write(Address(spender))
+			.write(value)
+			.write(nonce)
+			.write(deadline)
+			.build();
+		let permit_content = keccak_256(&permit_content);
+
+		let mut pre_digest = Vec::with_capacity(2 + 32 + 32);
+		pre_digest.extend_from_slice(b"\x19\x01");
```

Comment:
What does 0x19 refer to?

Reply:
See EIP712 (signed typed structures) : https://eips.ethereum.org/EIPS/eip-712#specification

Reply:
thanks!